### PR TITLE
Invalid starting communities

### LINKED ISSUE #29: "Invalid" starting communities fails
Our Leiden implementation allows you to provide a starting dictionary of node -> communities. 

The problem that @JonathanLarson found is that if you provide a community dictionary with nodes that have no direct edge relationship to anything in that community, the algorithm as-is falls apart. Leiden's enhancement to Louvain guarantees that partitions/clusters will only be made up of nodes that have some connection to each other - meaning that every node in that partition MUST contain an edge to another node in that partition - with the noted exception of clusters with only a single node.

We presumed that this would always be true within our leiden impl, but we didn't guarantee it if a user provided us a starting community mapping.  By happenstance @JonathanLarson was running leiden repeatedly over monthly snapshots of nodes and using the previous month's partitions as a starting point for generating the new partitions.  We ran into the one and only `panic! at the macro` in graspologic native because while two nodes were in this partition, they did not share an edge, and thus the subnetwork was "empty".

### PR DESCRIPTION
**Pull request by daxpryce**
Fixes #29

These already made it into `dev` at one point, but I rolled them back after I was sure we had found the culprit, since we couldn't repro on our own.

Now that I have and have written some unit tests, I think it's ready for an actual review.

### CODE REVIEW

File: packages/network_partitions/src/leiden/leiden.rs [hunk 1]
```diff
@@ -246,4 +316,36 @@ mod tests {
...
     }
+
+    #[test]
+    fn test_guarantee_clustering_sanity() {
+        let edges = edge_list();
+        let mut builder: LabeledNetworkBuilder<String> = LabeledNetworkBuilder::new();
+        let labeled_network: LabeledNetwork<String> = builder.build(edges.into_iter(), true);
+        let compact_network: &CompactNetwork = labeled_network.compact();
+        let mut clustering: Clustering = Clustering::as_self_clusters(compact_network.num_nodes());
+        // node 'a' and node 'f' do not share an edge
```
Comment (pbourke) [inline packages/network_partitions/src/leiden/leiden.rs:327]: comment says 'f', code says 'h'

File: packages/network_partitions/src/leiden/leiden.rs [hunk 2]
```diff
@@ -81,6 +82,13 @@ where
 
     let mut clustering: Clustering =
         clustering.unwrap_or(Clustering::as_self_clusters(network.num_nodes().clone()));
+
+    guarantee_clustering_sanity(network, &mut clustering)?;
+
+    // verify initial clustering provided is in a sane state for leiden to operate
+    // any node in a cluster must either be a singleton in that cluster or be connected to at least
+    // one other node in that cluster
```
Comment (jonmclean) [inline packages/network_partitions/src/leiden/leiden.rs:90]: Should the comment be before the guarantee_stuff method call?

File: packages/network_partitions/src/leiden/leiden.rs [hunk 3]
```diff
@@ -230,9 +238,71 @@ fn initial_clustering_for_induced(
...
+    for (_cluster, cluster_members) in &cluster_membership {
+        if cluster_members.len() > 1 {
+            // we are only trying to move non-singletons if they don't have a possible connection
+            for cluster_member in cluster_members {
+                let neighbors = node_neighbors.get(cluster_member).unwrap();
+                if neighbors.is_disjoint(cluster_members) {
+                    // we have no reason to be in this partition, because we have no links to anyone
+                    // else in it. we should make our own partition, with ___ and ___.
+                    let new_cluster: ClusterId = clustering.next_cluster_id();
+                    clustering.update_cluster_at(*cluster_member, new_cluster)?;
```
Comment (jonmclean) [inline packages/network_partitions/src/leiden/leiden.rs:269]: Updating the clustering is a much better solution than throwing an error.

**Review by jonmclean — APPROVED** [review]
Looks good

**Pull request merged**